server/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java

```diff
@@ -163,6 +163,25 @@ public Set<String> changedCustomProjectMetadataSet() {
         return result;
     }
 
+    /**
+     * Checks whether custom metadata type for a project has changed between the previous cluster state
+     * and the new cluster state. Custom metadata types are considered changed iff they have been added,
+     * updated or removed between the previous and the current state
+     */
+    public boolean customMetadataChanged(ProjectId projectId, String customMetadataType) {
+        Set<String> changed = new HashSet<>();
+        ProjectMetadata project = state.metadata().projects().get(projectId);
+        ProjectMetadata previousProject = previousState.metadata().projects().get(projectId);
+        if (previousProject != null && project != null) {
+            changed.addAll(changedCustoms(project.customs(), previousProject.customs()));
+        } else if (previousProject != null) {
+            changed.addAll(previousProject.customs().keySet());
+        } else if (project != null) {
+            changed.addAll(project.customs().keySet());
+        }
+        return changed.contains(customMetadataType);
```

Comment:
This still computes the entire `Set` unnecessarily. What do you think about something like this:
```suggestion
        ProjectMetadata previousProject = previousState.metadata().projects().get(projectId);
        ProjectMetadata project = state.metadata().projects().get(projectId);
        Object previousValue = previousProject == null ? null : previousProject.customs().get(customMetadataType);
        Object value = project == null ? null : project.customs().get(customMetadataType);
        return Objects.equals(previousValue, value) == false;
```